### PR TITLE
Optimize using the Multi-stage build feature.

### PR DESCRIPTION
**Pull request by jcouture**
## 📖 Description and motivation

Leaning in the Multi-stage build feature of Docker to heavily optimize the final container image. It no longer contains development dependencies or build artifacts. This has a two-fold effect of decreasing the size of the container image, but also reducing the potential security vulnerabilities.

#### Tasks

- [x] Split the original Dockerfile into Base, Build and Release stages:
- [x] The Base stage installs the bare minimum required for the project to run.
- [x] The Build stage takes care of installing the Ruby Gems, NPM Packages and pre-compiling the assets.
- [x] The Release stage inherits from the Base stage and copies the required files from the Build stage, leaving behind the build artifacts.

## 🎉 Result

We’ve observed as much as a 75% reduction of the final container image size.

## 🦀 Dispatch

`#dispatch/devops`

### CODE REVIEW

File: .dockerignore [hunk 1]
```diff
@@ -2,8 +2,10 @@ tmp/
 log/
 .git/
 node_modules/
-spec/
+spec
 .*
+*.md
+*.yaml
+*.yml
```
Comment (JoeyBG) [inline .dockerignore:9]: Have you tested this? I would have thought we needed it for https://github.com/mirego/killswitch/blob/main/config/secrets.yml and other config files

File: Dockerfile [hunk 2]
```diff
@@ -1,33 +1,58 @@
-FROM ruby:2.7.7-alpine3.16
+######################
+##### Base Stage #####
+######################
```
Comment (remi) [inline Dockerfile:3]: These are redundant I think 😉

File: Dockerfile [hunk 3]
```diff
@@ -1,33 +1,58 @@
...
 # Copy entrypoint script
 COPY scripts/docker-entrypoint.sh /usr/local/bin
 RUN chmod a+x /usr/local/bin/docker-entrypoint.sh
 
 # Create user
-RUN adduser -D killswitch
-RUN chown -R killswitch: /opt/killswitch
-USER killswitch
+RUN adduser -D -h /opt/killswitch -u 5000 killswitch && \
+  chown -R killswitch: /opt/killswitch
```
Comment (remi) [inline Dockerfile:51]: Is there a reason to join `adduser` and `chown` in a single `RUN` command here?

File: .dockerignore (hunk 1 above)
Comment (jcouture) [reply .dockerignore:9]: Good catch, reverting.

File: Dockerfile (hunk 2 above)
Comment (jcouture) [reply Dockerfile:3]: Got it. 😜

File: Dockerfile (hunk 3 above)
Comment (jcouture) [reply Dockerfile:51]: Each `RUN` instructions create an additional layer. Piping multiple commands to keep the image size as small as possible.

Reference: https://docs.docker.com/develop/develop-images/dockerfile_best-practices/#minimize-the-number-of-layers

File: Dockerfile [hunk 4]
```diff
@@ -1,33 +1,48 @@
...
+# Update system and install dependencies
+RUN apk --update --no-cache add --virtual build-dependencies build-base git nodejs npm postgresql-dev && \
+  bundle config set --local without 'development test' && \
+  bundle install && \
+  bundle binstubs --all && \
+  npm set progress=false && \
+  npm install --silent --production
+
+# Pre-compile assets
+RUN SECRET_KEY_BASE=__UNUSED_BUT_REQUIRED__ RAILS_ENV=production bundle exec rake assets:precompile
```
[the remark this answers is not in the document]
Comment (jcouture) [reply Dockerfile:25]: Great catch, of course we should! 🥳

**Review by remi — APPROVED** [review]

**Pull request merged**